ast/file_info.go

```diff
@@ -143,10 +143,16 @@ func (f *FileInfo) AddComment(comment, attributedTo Token) Comment {
 	}
 }
 
+// NodeInfo returns details from the original source for the given AST node. If
+// the given n is not a node from the AST corresponding to f, the returned value
+// is undefined (and interacting it with may incur panics).
```

Comment:
Not really. A node doesn't contain a backlink up to its parent, which means there's no way to traverse up to the root to see if it's associated with the same `*FileInfo`. Similarly, there's no link from the `*FileInfo` to its corresponding `*FileNode`. (Even if there were, it would be extremely inefficient to have search through the tree to find `n` to validate.)

FWIW, this is all an intentional trade-off for memory usage. If you look at the old compiler's [AST package](https://pkg.go.dev/github.com/jhump/protoreflect@v1.13.0/desc/protoparse/ast), its representation of node info (source positions, whitespace) was **way** simpler: every node had its [own info](https://github.com/jhump/protoreflect/blob/v1.13.0/desc/protoparse/ast/node.go#L84-L90). But this caused the memory usage of the AST to be bloated. For the new compiler, I took inspiration from how Go's [`go/parser`](https://pkg.go.dev/go/parser) and [`go/token`](https://pkg.go.dev/go/parser) packages work. The relationships between `ast.Node`, `ast.NodeInfo`, and `ast.FileInfo` in this package are roughly modeled after the relationships of `token.Pos`, `token.Position`, and `token.FileSet` in those packages. So this complexity was an intentional choice to reduce the memory footprint of the compiler.

We could possibly do a check to make sure that the node's `End()` isn't out of range in this file's set of tokens. But that can only be best effort: it would likely remove the likelihood of a panic, but the behavior is still bad/undefined if you use the wrong file and the node is _in_ range. Honestly, I don't think the check is worth it since we can't actually provide well-defined behavior.